cps/src/main/java/org/jenkinsci/plugins/workflow/cps/replay/ReplayAction.java

```diff
@@ -0,0 +1,254 @@
+    }
+
+    private @CheckForNull CpsFlowExecution getExecution() {
+        FlowExecutionOwner owner = ((FlowExecutionOwner.Executable) run).asFlowExecutionOwner();
+        if (owner == null) {
+            return null;
+        }
+        FlowExecution exec = owner.getOrNull();
+        return exec instanceof CpsFlowExecution ? (CpsFlowExecution) exec : null;
+    }
+
+    /* accessible to Jelly */ public boolean isEnabled() {
+        if (!run.hasPermission(REPLAY)) {
+            return false;
+        }
+        CpsFlowExecution exec = getExecution();
+        if (exec == null) {
+            return false;
+        }
+        if (exec.isSandbox()) {
+            return true;
+        } else {
+            // Whole-script approval mode. Can we submit an arbitrary script right here?
+            return Jenkins.getActiveInstance().hasPermission(Jenkins.RUN_SCRIPTS);
+        }
+    }
+
+    /** @see CpsFlowExecution#getScript */
+    /* accessible to Jelly */ public String getOriginalScript() {
+        CpsFlowExecution execution = getExecution();
+        return execution != null ? execution.getScript() : "???";
+    }
+
+    /** @see CpsFlowExecution#getLoadedScripts */
+    /* accessible to Jelly */ public Map<String,String> getOriginalLoadedScripts() {
+        CpsFlowExecution execution = getExecution();
+        return execution != null ? execution.getLoadedScripts() : /* ? */Collections.<String,String>emptyMap();
+    }
+
+    /* accessible to Jelly */ public Run getOwner() {
+        return run;
+    }
+
+    @Restricted(DoNotUse.class)
+    @RequirePOST
+    public void doRun(StaplerRequest req, StaplerResponse rsp) throws ServletException, IOException {
+        if (!isEnabled()) {
+            throw new AccessDeniedException("not allowed to replay"); // AccessDeniedException2 requires us to look up the specific Permission
+        }
+        JSONObject form = req.getSubmittedForm();
+        Map<String,String> otherScripts = new HashMap<String,String>();
+        for (Map.Entry<String,String> entry : getOriginalLoadedScripts().entrySet()) {
+            otherScripts.put(entry.getKey(), form.optString(entry.getKey(), entry.getValue()));
+        }
+        run(form.getString("mainScript"), otherScripts);
+        rsp.sendRedirect("../.."); // back to WorkflowJob; new build might not start instantly so cannot redirect to it
+    }
+
+    private static final Iterable<Class<? extends Action>> COPIED_ACTIONS = ImmutableList.of(
+        ParametersAction.class,
+        SCMRevisionAction.class
+    );
+
+    /** For whitebox testing. */
+    public @CheckForNull QueueTaskFuture/*<Run>*/ run(String script) {
+        return run(script, Collections.<String,String>emptyMap());
+    }
+
+    /**
+     * For whitebox testing.
+     * @param script main script; replacement for {@link #getOriginalScript}
+     * @param otherScripts auxiliary scripts, keyed by class name; replacement for {@link #getOriginalLoadedScripts}
```

Comment:
I feel like "otherScripts" is uninformative & confusing enough a name to merit a :bug:  here.  I'd suggest this should be "replacementScripts"